sktime/classification/dictionary_based/_boss.py

```diff
@@ -406,10 +409,13 @@ def __init__(self,
         super(BOSSIndividual, self).__init__()
 
     def fit(self, X, y):
-        X, y = check_X_y(X, y, enforce_univariate=True)
+
+        if isinstance(X, pd.Series) or isinstance(X, pd.DataFrame):
+            X, y = check_X_y(X, y, enforce_univariate=True)
+            X = tabularize(X, return_array=True)
 
         sfa = self.transformer.fit_transform(X)
-        self.transformed_data = sfa.iloc[:, 0]
+        self.transformed_data = sfa[0]  # .iloc[:, 0]
```

Comment:
Hi @mloning, 

maybe leave the config as it was? The dependency to "black" requires reformatting all existing source codes. I am not sure, if that is intended? Most likely this makes spotting differences in pull requests to the former code even harder?

Very good spot on regarding .iloc :) I have noticed a general problem in sktime which significantly reduces runtimes and memory consumption for dictionary-methods. 

For example in BOSS we have:
BOSSEnsemble -> BOSSIndividual -> SFA. And each one requires a Pandas Dataframe as input, next converts the DataFrame using tabularize to a Numpy Array. In the end a Pandas Dataframe is returned (converted from array or dict). These conversions were very expensive for SFA and BOSSIndivudual as they consumed a significant proportion of the runtime and memory. Thus, I have changed the internal dataflow such that "tabularize" is only performed in BOSSEnsemble and numpy arrays are passed down to BOSSIndividual and SFA, saving all conversion steps inbetween. 

I noticed similar issues with PAA, but didnt address these (I suppose a lot of time from ShapeletTransform is due to the conversion steps for PAA). Maybe a more general approach would be desirable here (e.g. allowing numpy or pandas input for the feature extractors)?

Do you have an example for adding lines to CODEOWNERS? There is no entry yet...

Thank you!

Reply:
Hi @patrickzib 
* Yes, the data container is an ongoing discussion point! I believe we made a bad choice at the start in going for nested pandas data frames instead of numpy arrays. Ideally this would be solved in the input check functions `check_X` and `check_X_y`, we're also thinking about adding support for 3d numpy arrays with shape = [n_instances, n_variables, n_timepoints] to avoid the conversion altogether for equal length series. What are your thoughts on this? How do you represent unequal length data in Python with varying number of observations across instances and/or variables?  
* CI is set up to only check the files that were changed, so this shouldn't be an issue for any new changes after pre-commit was set up. So we can ignore the failing tests on this PR. I see the point about spotting differences in PRs when additional enforcing new formatting rules. Mh, will think about removing black again, but so can you just undo the change to setup.cfg? 
* You can find more info on CODEOWNERS here: https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners

Reply:
Does that mean that SFA's transform method now returns numpy arrays? If so, wouldn't this break other pipelines where the next steps expects pandas data frames?

Reply:
Or does SFA return unequal length data? I'd be very interested in hearing your opinion on the data container issue! Perhaps a quick call would be easiest?

Reply:
@mloning Thank you for your answers! 

I have a 3 days old baby at home, thus it is a bit difficult to plan my time ahead of time :)

- It would be great if check_X and check_X_y would convert the data into an input format that can be used within sklearn pipelines. It would also be great if the format is compatible with numba, which supports numpy but not pandas, as of now.

- I did a kind of workaround in this pull request to avoid breaking existing functionality.
1.  For the return type of the methods: There is a constructor parameter called: `return_pandas_data_series`. When set to `False`, it will return numpy arrays. When set to `True`, it will return the default in sktime (pandas.dataframes and series). Thus, I can choose which output to use by initializing the classes, e.g. SFA(return_pandas_data_series=True) when pandas output is needed. 

2. For the input to the methods: The `fit` and `transform` methods check the input format of X and y and call `tabularize()` if it is in pandas format or keep the numpy arrays, if already in the correct format. But it would be great, if check_X and check_X_y would in the long run call tabularize directly.

- I do not really have an answer for variable lengths in numpy :(

What do you think?

Reply:
Oh wow, congratulations! :) 

I agree, the solution isn't ideal, doing it through the input checks/conversion would be better, but don't mind it as a temporary fix. We're considering awkward-array as an alternative to numpy for unequal length data.

Reply:
Thank you :) very exciting times...

awkward-array looks promissing! There is even an extension to numba, too: https://pypi.org/project/awkward-numba/